### PR TITLE
setup improvements

### PR DESCRIPTION
**Pull request by dougsland**
- demo: validate if commands exists
- e2e: add sudo command to template file
- getopt(), now users can use --installdir and --rootfs
- --help and usage message
- added root_check()
- constants
- use of bash to include array type
- demo: rename asilservice.py

### CODE REVIEW

**Comment by dougsland** [comment]
Just a feedback, in my system I noticed the qm failure due missing `/run/asilservice`, that's the reason for the renaming.

Output:
```
Jul 25 04:47:37 node1 systemd[1]: Failed to start qm.service.
Jul 25 04:47:37 node1 systemd[1]: qm.service: Failed with result 'exit-code'.
Jul 25 04:47:37 node1 systemd[1]: var-lib-containers-storage-overlay.mount: Deactivated successfully.
Jul 25 04:47:37 node1 systemd[1]: qm.service: Main process exited, code=exited, status=125/n/a
Jul 25 04:47:37 node1 systemd[1]: var-lib-containers-storage-overlay.mount: Deactivated successfully.
Jul 25 04:47:37 node1 qm[22805]: Error: statfs /run/asilservice: no such file or directory
```

I will take this week to dig more the setup and fix issues raised.

**Comment by dougsland** [comment]
Showing the new setup:

```
# ./setup --help
The setup script will install an OS environment by default into
/usr/lib/qm/rootfs and create a Podman quadlet containerized environment
running systemd as PID1 inside.

Usage: setup [OPTIONS]

OPTIONS:

--help        Print this message.
--installdir  qm install directory (default: /usr/share/qm)
--rootfs      set rootfs (default: /usr/lib/qm/rootfs)

Example:
  $ sudo ./setup --installdir=/usr/share/qm --rootfs=/usr/lib/qm/rootfs
```


```
# ./setup
setup has started...
  * rootfs: /usr/lib/qm/rootfs
  * install dir: /usr/share/qm

Last metadata expiration check: 1:24:23 ago on Tue Jul 25 03:45:46 2023.
Package selinux-policy-targeted-38.1.17-1.el9.noarch is already installed.
Package podman-101:4.5.2~dev-1.20230706090955027646.v4.5.8.g813f1b53b.el9.x86_64 is already installed.
Package systemd-252-15.el9.x86_64 is already installed.
Package hirte-agent-0.5.0-0.202307241429.gitfadad7d.el9.x86_64 is already installed.
Package procps-ng-3.3.17-11.el9.x86_64 is already installed.
Dependencies resolved.
Nothing to do.
Complete!
Last metadata expiration check: 1:24:24 ago on Tue Jul 25 03:45:46 2023.
Dependencies resolved.
Nothing to do.
Complete!
```

```
# ./setup --installdir=bar --rootfs=foo
setup has started...
  * rootfs: foo
  * install dir: bar

Config error: Error parsing 'foo': given path 'foo' is not absolute.
```

File: demos/demo [hunk 1]
```diff
@@ -23,7 +23,16 @@ echo_color() {
...
+check_command() {
+    if ! command -v "${1}" &> /dev/null
+    then
+        echo "${1} command could not be found. Install ${1} to continue."
+        exit 1
+    fi
+}
+
 init() {
+    check_command "sudo"
```
Comment (sandrobonazzola) [inline demos/demo:35]: Is it really worth it? just checking `sudo` but not checking the other commands?

**Review by sandrobonazzola — APPROVED** [review]
LGTM. Just a note about checking `sudo` but not the other commands.

File: demos/demo (hunk 1 above)
Comment (dougsland) [reply demos/demo:35]: good catch, now that we have bash, let's use a for there to validate the commands. I just used sudo as I noticed e2e had no sudo and failed.

**Comment by dougsland** [comment]
CI error not related, it's something else. A trivial change also triggered the same issue: https://artifacts.dev.testing-farm.io/664e4f4f-d2ca-41eb-b8e7-3cce25df56c7/

**Comment by dougsland** [comment]
I work to fix this issue in a different patch.

File: setup [hunk 2]
```diff
@@ -69,14 +93,55 @@ install() {
...
+  -- "$@"
+)
+
+eval set --"${opts}"
+
+# main()
+root_check
+
+echo "setup has started..."
+echo  "  * rootfs: ${ROOTFS}"
```
Comment (rhatdan) [inline setup]: Shouldn't this happen after the getopt loop below?

Comment (dougsland) [reply setup]: In one of the patches below, I changed that. Sorry,if you prefer I can do a squash.

File: setup [hunk 3]
```diff
@@ -142,6 +137,11 @@ while [[ $# -gt 0 ]]; do
 done
 
 # Figure out the options, execute setup
+echo "setup has started..."
```
Comment (rhatdan) [inline setup:140]: Ok you did shift these. can you merge this one with the getopt PR?

Comment (dougsland) [reply setup:140]: sure

**Comment by rhatdan** [comment]
LGTM but it would be nice to add some tests of the options.

**Comment by dougsland** [comment]
> LGTM but it would be nice to add some tests of the options.

Agreed. More patches to come. Let me fix first CI.

**Pull request merged**